### PR TITLE
Docker based Jupyter Reports 

### LINKED ISSUE #3471: Jupyter Support - Docker and Module report support
Spec: https://docs.google.com/document/d/1UBjMaaeA2ZrYbjAcdou8gX6fQK84LfmzRCjmMAkI3oM/edit?usp=sharing
Remaining work from scrumwise: https://www.scrumwise.com/scrum/#/backlog-item/10424-labkey-rd-enterprise-improvements-jupyter-reports-ipynbreport-part-2/id-1810-46784-55

- [x] inject helper code into jupyter report (3) 
- [x] get reviews from pythonistas and move into labkey python package (3)
- [x] write helper to read report_config.json and environment variables (3)
- [x] race condition (1)
- [x]  Ramp-up (5) - @labkey-tchad
- [x]  Ramp-up (5) - @XingY 
- [x] code review (2) - @XingY 
- [x]  Manual testing (5) - @labkey-tchad
- [x]  Teamcity review and merge (1) - @labkey-klum 
- [x] Docs handoff (1) - @labkey-klum

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
Docker based approach to run Jupyter reports and display the results in LabKey server. This PR contains the development work for:
- Jupyter (ipynb) report type
- execute the report in a docker image and push/pull results into/out of the docker container
- surface a new docker scripting engine to run the reports
- integration of Jupyter reports into the script editor, reporting framework etc.

[specification](https://docs.google.com/document/d/1UBjMaaeA2ZrYbjAcdou8gX6fQK84LfmzRCjmMAkI3oM/edit)

#### Related Pull Requests
https://github.com/LabKey/platform/pull/3474
https://github.com/LabKey/docker/pull/71

### CODE REVIEW

File: api/src/org/labkey/api/docker/DockerService.java [hunk 1]
```diff
@@ -450,6 +458,8 @@ public DockerContainer(
 
     DockerContainer start(ImageConfig image, String prefix, User user, Map<String, String> labels, Map<String, String> env, Map<File, String> filesForContainer, Map<InputStream, String> streamsForContainer, List<List<String>> postStartCmds, ContainerUsage usage) throws IOException;
 
+    RunResult run(ImageConfig image, String prefix, Map<String, String> envAdditional, InputStream stdin, OutputStream stdout, OutputStream stderr);
```
Comment (XingY) [inline api/src/org/labkey/api/docker/DockerService.java]: update stdin to tarInputStream

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+                catch (IOException ex)
+                {
+                    bgException[0] = new DbScope.RetryPassthroughException(ex);
+                }
+            });
+            t.start();
+
+            String tempDir = "/tmp/" + GUID.makeGUID();
+            var environment = Map.of(
+                    "TEMP_DIRECTORY", tempDir,
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java]: I don't think this "TEMP_DIRECTORY is used by the script. Though entrypoint creates this directory, the actual tmp directory used is created in exec_runreport_tar using `$(mktemp -d)`

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+                return inputScript;
+            return null;
+        }
+    }
+
+
+    class DockerRunTarStdinStdout implements ExecuteStrategy
+    {
+        File inputScript;
+        File outputDocument;
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java:339]: I don't believe outputDocument is ever used?

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+        }
+
+        @Override
+        public int execute(ViewContext context, String apiKey, File working, File ipynb) throws IOException
+        {
+            inputScript = ipynb;
+
+            JSONObject reportConfig = createReportConfig(context, ipynb);
+            // I tried "putting" a fake tar entry, but TarArchiveOutputStream seems to actually want the file to exist
+            FileUtils.write(new File(working,"report_config.json"), reportConfig.toString(), StringUtilsLabKey.DEFAULT_CHARSET);
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java]: Perhaps report_config.json should be a environment var rather than hard coded?

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+                    bgException[0] = new DbScope.RetryPassthroughException(ex);
+                }
+            });
+            t.start();
+
+            String tempDir = "/tmp/" + GUID.makeGUID();
+            var environment = Map.of(
+                    "TEMP_DIRECTORY", tempDir,
+                    "LABKEY_API_KEY", apiKey);
+            try (var run = DockerService.get().run(image, "ipynb", environment, in, out, err))
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java]: EXTENSION const instead of "ipynb"

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+            return DockerService.get().getImageConfigBuilder(image.getImageName()).build();
+        }
+        else
+        {
+            throw new IllegalStateException("No script engine configured for  " + LABEL + " reports");
+        }
+    }
+
+
+    class DockerRunIpynbStdinStdout implements ExecuteStrategy
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java]: when will this be used? Was this experimental or is it still needed?

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+
+
+//    final StringBuilder stdout = new StringBuilder();
+//    final StringBuilder stderr = new StringBuilder();
+//   new Thread(() -> Readers.getReader(process.getInputStream()).lines().forEach(stdout::append)).start();
+//   new Thread(() -> Readers.getReader(process.getErrorStream()).lines().forEach(stderr::append)).start();
+//
+
+    // process that writes the computed ipynb file to stdout
+    class ExecIpynbStdout implements ExecuteStrategy
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java:253]: Same question on if this is it still needed?

File: api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+                }
+                sb.append(HtmlString.unsafe("</div>")); // ipynb-cell
+            }
+            sb.append(HtmlString.unsafe("</div>")); // labkey-wiki
+            return sb.toString();
+        }
+
+        private void renderStylesheet(HtmlStringBuilder sb)
+        {
+            sb.append(HtmlString.unsafe(
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java:239]: Maybe move this to a css selector?

File: api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+                    return;
+                }
+                /* ERROR
+                {
+                  "ename": "KeyError",
+                  "output_type": "error",
+                  "evalue": "'domain'",
+                  "traceback": [
+                    "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+                    "\u001b[0;31mKeyError\u001b[0m Traceback (most recent call last)",
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java]: remove this block?

File: core/src/org/labkey/core/view/configReportsAndScripts.jsp [hunk 10]
```diff
@@ -62,873 +61,939 @@
...
+        var PERL_EXTENSIONS = 'pl';
+        var R_ENGINE_NAME = 'R Scripting Engine';
+        var REMOTE_R_ENGINE_NAME = 'Remote R Scripting Engine';
+        var R_DOCKER_ENGINE_NAME = 'R Docker Scripting Engine';
+        var DOCKER_REPORT_NAME = 'Docker Report Engine';
+        var defaultR, countR = 0;
+        let grid;
+        let remoteEnabled = <%=isRemoteEnabled%>;
+        let dockerServiceEnabled = <%=isDockerServiceEnabled%>;
+        let rDockerEnabled = true;//<%=isRDockerAvailable%>;
```
Comment (XingY) [inline core/src/org/labkey/core/view/configReportsAndScripts.jsp]: always true? shouldn't this also check for dockerService?

File: core/src/org/labkey/core/view/configReportsAndScripts.jsp [hunk 11]
```diff
@@ -62,873 +61,939 @@
...
+            // hostReadOnlyMount: {label: 'Mount (ro): host directory', description: "Additional mount: host directory (read-only). Optional read-only mount point"},
+            // containerReadOnlyMount: {label: 'Mount (ro): container directory', description: "Additional mount: container directory (read-only). Optional read-only mount point"},
+            // hostReadWriteMount: {label: 'Mount (rw): host directory', description: "Additional mount: host directory (read and write). Optional read/write mount point"},
+            // containerReadWriteMount: {label: 'Mount (rw): container directory', description: "Additional mount: container directory (read and write). Optional read/write mount point"},
+            // appArmorProfile: {label: 'AppArmor Profile', description: ''},
+            extraENVs: {label: 'Extra Variables', description: 'Additional environment variables to be passed in when running a container. Usage example: &apos;OPTION1=1000,OPTION2=red&apos;, which will be converted to &apos;-e OPTION1=1000 -e OPTION2=red&apos; for docker run command.'}
```
Comment (XingY) [inline core/src/org/labkey/core/view/configReportsAndScripts.jsp]: It doesn't look like this is wired up for ipynb

File: query/webapp/reports/ScriptReportPanel.js [hunk 12]
```diff
@@ -462,6 +490,61 @@ Ext4.define('LABKEY.ext4.ScriptReportPanel', {
             });
         }
 
+        if (this.reportConfig.jupyterOptions) {
```
Comment (XingY) [inline query/webapp/reports/ScriptReportPanel.js]: Consider extracting jupyter specific into a separate file, see how ExternalEditorHelper.js for RStudio example

File: query/webapp/scriptreporteditor.lib.xml [hunk 13]
```diff
@@ -1,5 +1,5 @@
 <libraries xmlns="http://labkey.org/clientLibrary/xml/">
-    <library compileInProductionMode="true">
+    <library compileInProductionMode="false">
```
Comment (XingY) [inline query/webapp/scriptreporteditor.lib.xml:2]: curious why is this change needed

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+    public String getDefaultScript()
+    {
+        return """
+                {
+                  "cells": [
+                    {
+                      "cell_type": "code",
+                      "source": [
+                        "from ReportConfig import get_report_api_wrapper, get_report_data, get_report_parameters\\n",
+                        "print(get_report_parameters())\\n",
```
Comment (XingY) [inline api/src/org/labkey/api/reports/report/python/IpynbReport.java:127]: I'm getting `TypeError: 'NoneType' object is not subscriptable` error on `print(get_report_parameters())`, but maybe that's just me? @labkey-tchad

**Review by XingY — COMMENTED** [review]
That is quite a lot of code... To be honest I didn't read through all of it, but glanced at least most of it. Let me know if there are specific files that needs more eyes on.

Thanks @XingY, I can help with some of the new report code but if you could focus on the changes to the docker specific code (I'm less familiar with that).

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 14 above)
Comment (labkey-tchad) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java:127]: How long since you updated your docker image? I think I fixed that

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 6 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java]: good question, @labkey-matthewb any thoughts?

**Comment by labkey-tchad** [comment]
The EHR BVTs have been failing in this branch ever since updating to docker-java 3.2.8:
```
  java.lang.LinkageError: loader constraint violation: when resolving method 'void javax.mail.BodyPart.setDataHandler(javax.activation.DataHandler)' the class loader org.labkey.bootstrap.LabKeyBootstrapClassLoader @6a4f1a55 of the current class, org/labkey/core/notification/EmailServiceImpl$EmailMessageImpl, and the class loader java.net.URLClassLoader @2e817b38 for the method's defining class, javax/mail/Part, have different Class objects for the type javax/activation/DataHandler used in the signature (org.labkey.core.notification.EmailServiceImpl$EmailMessageImpl is in unnamed module of loader org.labkey.bootstrap.LabKeyBootstrapClassLoader @6a4f1a55, parent loader java.net.URLClassLoader @2e817b38; javax.mail.Part is in unnamed module of loader java.net.URLClassLoader @2e817b38, parent loader 'app')
  at org.labkey.core.notification.EmailServiceImpl$EmailMessageImpl.createMessage(EmailServiceImpl.java:314)
  at org.labkey.core.notification.EmailServiceImpl$TestCase.convertMessageToString(EmailServiceImpl.java:396)
  at org.labkey.core.notification.EmailServiceImpl$TestCase.testEmailAttachments(EmailServiceImpl.java:349)
```

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java]: That is correct.  This is not currently used, but I was wondering if it would be useful when/if we reuse containers potentially concurrently running reports.  Should comment or remove.

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 3 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java:339]: I imagined that the more general version of Docker report would include a report_output.json which would declare the return artifacts.  This is not being used here because there is only one artifact.  Should comment the intention, and remove the unused variable.

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 4 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java]: That's a good idea (with corresponding change to ReportConfig.py)

File: query/webapp/scriptreporteditor.lib.xml (hunk 13 above)
Comment (labkey-klum) [reply query/webapp/scriptreporteditor.lib.xml:2]: The script editor uses a `FileReader` object to import a file into the editor, it's not compatibly with the version of JS in the tool to minify the libraries. I have it on my list to make a decision to leave it this way, replace yuicompressor, or use a different solution to import files.

File: core/src/org/labkey/core/view/configReportsAndScripts.jsp (hunk 10 above)
Comment (labkey-klum) [reply core/src/org/labkey/core/view/configReportsAndScripts.jsp]: good catch, was debugging something

File: core/src/org/labkey/core/view/configReportsAndScripts.jsp (hunk 11 above)
Comment (labkey-klum) [reply core/src/org/labkey/core/view/configReportsAndScripts.jsp]: true, I'll remove it

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 14 above)
Comment (XingY) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java:127]: yep, it's working now with the latest image. thanks

File: query/webapp/reports/ScriptReportPanel.js (hunk 12 above)
Comment (labkey-klum) [reply query/webapp/reports/ScriptReportPanel.js]: sure, it's a little different from the way editor options are handled but I like moving the report specific options out of the main file.

File: api/src/org/labkey/api/docker/DockerService.java (hunk 1 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/docker/DockerService.java]: 👍

**Comment by labkey-matthewb** [comment]
> javax

$ grep javax.activation.DataHandler -R * --include "*.jar"
grep: apache-tomcat-9.0.50/lib/javax.activation.jar: binary file matches
grep: build/staging/tomcat-lib/javax.activation.jar: binary file matches
grep: build/deploy/modules/docker/lib/jakarta.activation-api-1.2.1.jar: binary file matches
grep: build/modules/docker/explodedModule/lib/jakarta.activation-api-1.2.1.jar: binary file matches

What is the gradle magic to force docker-java to not pull "jakarta.activation-api-1.2.1.jar"?

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 2 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java]: removed for now

File: api/src/org/labkey/api/reports/report/python/IpynbReport.java (hunk 5 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/reports/report/python/IpynbReport.java]: agreed

File: api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java (hunk 9 above)
Comment (labkey-klum) [reply api/src/org/labkey/api/reports/report/r/view/IpynbOutput.java]: done

**Review by XingY — APPROVED** [review]

File: query/webapp/scriptreporteditor.lib.xml (hunk 13 above)
Comment (labkey-klum) [reply query/webapp/scriptreporteditor.lib.xml:2]: @labkey-nicka @cnathe thoughts on whether this is okay to do for the time being, other than increasing the payload on production builds are there other downsides to setting this flag false?
